src/cid.ts

```diff
@@ -0,0 +1,86 @@
+import {Logger} from './logging/index';
+import {formatCause} from './error';
+import NullLogger from './logging/nullLogger';
+
+export interface CidAssigner {
+    assignCid(): Promise<string>;
+}
+
+export class CachedAssigner implements CidAssigner {
+    private readonly generator: CidAssigner;
+
+    private readonly storage: Storage;
+
+    private readonly key: string;
+
+    private readonly logger: Logger;
+
+    public constructor(generator: CidAssigner, storage: Storage, key: string, logger?: Logger) {
+        this.generator = generator;
+        this.key = key;
+        this.storage = storage;
+        this.logger = logger ?? new NullLogger();
+    }
+
+    public async assignCid(): Promise<string> {
+        const cid = this.storage.getItem(this.key);
+
+        if (cid !== null) {
+            this.logger.debug('Previous CID loaded from cache');
+
+            return cid;
+        }
+
+        const newCid = await this.generator.assignCid();
+
+        this.logger.debug('New CID stored into cache');
+
+        this.storage.setItem(this.key, newCid);
```

Comment:
Being a little picky, shouldn't the log come _after_the `setItem`?
Similar to how the log for when it came from the cache comes after the `getItem` call

Reply:
Agreed